File: client/old/src/main/java/net/treset/minecraftlauncher/util/string/FormatString.java
package net.treset.minecraftlauncher.util.string;

public abstract class FormatString {
    public static int distance(String a, String b) {
        a = a.toLowerCase();
        b = b.toLowerCase();
        int [] costs = new int [b.length() + 1];
        for (int j = 0; j < costs.length; j++)
            costs[j] = j;
        for (int i = 1; i <= a.length(); i++) {
            costs[0] = i;
            int nw = i - 1;
            for (int j = 1; j <= b.length(); j++) {
                int cj = Math.min(1 + Math.min(costs[j], costs[j - 1]), a.charAt(i - 1) == b.charAt(j - 1) ? nw : nw + 1);
                nw = costs[j];
                costs[j] = cj;
            }
        }
        return costs[b.length()];
    }

    public abstract String get() throws FormatException;

    @Override
    public String toString() {
        try {
            return get();
        } catch (FormatException e) {
            return "";
        }
    }

    public static class FormatException extends Exception {
        public FormatException(String message) {
            super(message);
        }

        public FormatException(String message, Throwable cause) {
            super(message, cause);
        }
    }

}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/util/string/FormatString.kt
package net.treset.treelauncher.backend.util.string

import net.treset.treelauncher.backend.util.string.FormatString.FormatException
import java.util.*
import kotlin.math.min

abstract class FormatString {
    @Throws(FormatException::class)
    abstract fun get(): String
    override fun toString(): String {
        return try {
            get()
        } catch (e: FormatException) {
            ""
        }
    }

    class FormatException : Exception {
        constructor(message: String?) : super(message)
        constructor(message: String?, cause: Throwable?) : super(message, cause)
    }

    companion object {
        fun distance(a: String, b: String): Int {
            var a = a
            var b = b
            a = a.lowercase(Locale.getDefault())
            b = b.lowercase(Locale.getDefault())
            val costs = IntArray(b.length + 1)
            for (j in costs.indices) costs[j] = j
            for (i in 1..a.length) {
                costs[0] = i
                var nw = i - 1
                for (j in 1..b.length) {
                    val cj = min(
                        (1 + min(costs[j].toDouble(), costs[j - 1].toDouble())).toDouble(),
                        (if (a[i - 1] == b[j - 1]) nw else nw + 1).toDouble()
                    ).toInt()
                    nw = costs[j]
                    costs[j] = cj
                }
            }
            return costs[b.length]
        }
    }
}
